File: Packages/AgentGateway/Sources/AgentGateway/GatewayClient.swift
import Foundation
import os.log

private let log = Logger(subsystem: "com.priyanshu.metamod", category: "AgentGateway")

public struct GatewayConfig: Sendable, Equatable {
    public var host: String
    public var port: Int
    public var useTLS: Bool
    public var token: String

    public init(host: String = "127.0.0.1", port: Int = 18789, useTLS: Bool = false, token: String = "") {
        self.host = host
        self.port = port
        self.useTLS = useTLS
        self.token = token
    }

    public var url: URL? {
        var components = URLComponents()
        components.scheme = useTLS ? "wss" : "ws"
        components.host = host
        components.port = port
        if !token.isEmpty { components.queryItems = [URLQueryItem(name: "token", value: token)] }
        return components.url
    }
}

public enum GatewayConnectionState: Sendable, Equatable {
    case disconnected
    case connecting
    case connected
    case waitingForPairing
    case error(String)
}

/// Handles `node.invoke` commands forwarded by the gateway (camera.snap, device.status, …).
/// Main-actor isolated so non-Sendable JSON params stay on the actor.
@MainActor
public protocol NodeCommandHandler: AnyObject {
    /// Returns the result payload, or throws `GatewayError`.
    func handle(method: String, params: [String: Any]) async throws -> [String: Any]
}

/// WebSocket client for the OpenClaw/Hermes gateway protocol. Connects as a `node`,
/// signs the challenge with the device identity, and dispatches `node.invoke` to a handler.
/// The same client works for OpenClaw and Hermes — only the endpoint differs.
@MainActor
public final class GatewayClient: ObservableObject {
    @Published public private(set) var state: GatewayConnectionState = .disconnected

    private var config: GatewayConfig
    private let identity: DeviceIdentity
    private weak var handler: NodeCommandHandler?
    private let nowMs: @Sendable () -> Int64

    private var task: URLSessionWebSocketTask?
    private var session: URLSession
    private var keepaliveTask: Task<Void, Never>?

    private let clientID = "metamod-ios"
    private let scopes = ["operator.read", "operator.write"]

    public init(
        config: GatewayConfig,
        identity: DeviceIdentity,
        handler: NodeCommandHandler?,
        session: URLSession = .shared,
        nowMs: @escaping @Sendable () -> Int64 = { Int64(Date().timeIntervalSince1970 * 1000) }
    ) {
        self.config = config
        self.identity = identity
        self.handler = handler
        self.session = session
        self.nowMs = nowMs
    }

    public func update(config: GatewayConfig) { self.config = config }

    public func connect() {
        guard let url = config.url else { state = .error("Invalid gateway URL"); return }
        disconnect()
        state = .connecting
        let task = session.webSocketTask(with: url)
        self.task = task
        task.resume()
        receiveLoop()
    }

    public func disconnect() {
        keepaliveTask?.cancel(); keepaliveTask = nil
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
        state = .disconnected
    }

    // MARK: Receive

    private func receiveLoop() {
        task?.receive { [weak self] result in
            Task { @MainActor [weak self] in
                guard let self else { return }
                switch result {
                case let .success(message):
                    self.handleMessage(message)
                    self.receiveLoop()
                case let .failure(error):
                    self.state = .error(error.localizedDescription)
                }
            }
        }
    }

    private func handleMessage(_ message: URLSessionWebSocketTask.Message) {
        let data: Data
        switch message {
        case let .data(d): data = d
        case let .string(s): data = Data(s.utf8)
        @unknown default: return
        }
        guard let frame = GatewayFrame.decode(data) else { return }
        switch frame {
        case let .event(event, payload):
            if event == "connect.challenge", let nonce = payload["nonce"] as? String {
                sendConnect(nonce: nonce)
            }
        case let .response(_, ok, payload, error):
            if let error {
                state = error.code == "NOT_PAIRED" ? .waitingForPairing : .error(error.message)
            } else if ok, case .connecting = state {
                state = .connected
                startKeepalive()
                _ = payload
            }
        case let .request(id, method, params):
            dispatch(id: id, method: method, params: params)
        }
    }

    // MARK: Connect handshake

    private func sendConnect(nonce: String) {
        let signedAt = nowMs()
        let signedString = GatewayAuth.signedString(
            deviceID: identity.deviceID, clientID: clientID, clientMode: "node",
            role: "node", scopes: scopes, signedAtMs: signedAt,
            token: config.token, nonce: nonce, platform: "ios", deviceFamily: "rayban"
        )
        let signature = (try? identity.sign(signedString)) ?? ""
        let params: [String: Any] = [
            "minProtocol": 3, "maxProtocol": 4,
            "client": ["id": clientID, "mode": "node", "name": "Meta-Mod"],
            "role": "node",
            "scopes": scopes,
            "caps": ["camera"],
            "commands": ["camera.snap", "camera.list", "device.status", "device.info"],
            "auth": ["token": config.token],
            "device": [
                "id": identity.deviceID,
                "publicKey": identity.publicKeyB64URL,
                "nonce": nonce,
                "signedAt": signedAt,
                "signature": signature,
                "platform": "ios",
                "deviceFamily": "rayban"
            ]
        ]
        send(.request(id: UUID().uuidString, method: "connect", params: params))
    }

    // MARK: node.invoke dispatch

    private func dispatch(id: String, method: String, params: [String: Any]) {
        guard method == "node.invoke" else {
            send(.response(id: id, ok: false, payload: nil, error: GatewayError(code: "UNKNOWN_METHOD", message: method)))
            return
        }
        let command = params["command"] as? String ?? ""
        let args = params["params"] as? [String: Any] ?? [:]
        Task { @MainActor [weak self] in
            guard let self, let handler = self.handler else {
                self?.send(.response(id: id, ok: false, payload: nil, error: GatewayError(code: "NO_HANDLER", message: "No node handler")))
                return
            }
            do {
                let payload = try await handler.handle(method: command, params: args)
                self.send(.response(id: id, ok: true, payload: payload, error: nil))
            } catch let error as GatewayError {
                self.send(.response(id: id, ok: false, payload: nil, error: error))
            } catch {
                self.send(.response(id: id, ok: false, payload: nil, error: GatewayError(code: "ERROR", message: error.localizedDescription)))
            }
        }
    }

    // MARK: Keepalive

    private func startKeepalive() {
        keepaliveTask?.cancel()
        keepaliveTask = Task { @MainActor [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 15 * 1_000_000_000)
                guard let self, self.state == .connected else { return }
                self.send(.request(id: UUID().uuidString, method: "ping", params: [:]))
            }
        }
    }

    // MARK: Send

    private func send(_ frame: GatewayFrame) {
        guard let data = try? frame.encoded(), let string = String(data: data, encoding: .utf8) else { return }
        task?.send(.string(string)) { error in
            if let error { log.error("send failed: \(error.localizedDescription)") }
        }
    }
}

File: android/src/gateway/GatewayClient.ts
import type { DeviceIdentity } from './identity';
import { decodeFrame, encodeFrame, Frame, GatewayError, signedStringV3 } from './protocol';

export interface GatewayConfig {
  host: string;
  port: number;
  useTLS: boolean;
  token: string;
}

export const DEFAULT_CONFIG: GatewayConfig = {
  host: '127.0.0.1',
  port: 18789,
  useTLS: false,
  token: '',
};

export function configURL(c: GatewayConfig): string {
  const scheme = c.useTLS ? 'wss' : 'ws';
  const base = `${scheme}://${c.host}:${c.port}`;
  return c.token ? `${base}?token=${encodeURIComponent(c.token)}` : base;
}

export type ConnectionState =
  | 'disconnected'
  | 'connecting'
  | 'connected'
  | 'waitingForPairing'
  | 'error';

// Returns the result payload, or throws { code, message }.
export type CommandHandler = (
  method: string,
  params: Record<string, any>,
) => Promise<Record<string, any>>;

// A line in the session's "working traces" feed.
export interface Trace {
  id: string;
  kind: 'system' | 'agent' | 'user' | 'error';
  text: string;
  detail?: string;
  at: number;
}

const CLIENT_ID = 'metamod-ios';
const SCOPES = ['operator.read', 'operator.write'];

// Faithful port of the native GatewayClient: connects as a `node`, answers the
// gateway's Ed25519 challenge, keeps alive with pings, and dispatches node.invoke.
export class GatewayClient {
  state: ConnectionState = 'disconnected';
  lastError = '';

  private ws: WebSocket | null = null;
  private keepalive: ReturnType<typeof setInterval> | null = null;

  constructor(
    private config: GatewayConfig,
    private identity: DeviceIdentity,
    private handler: CommandHandler,
    private onState: (state: ConnectionState, error: string) => void,
    private onTrace?: (t: Trace) => void,
  ) {}

  update(config: GatewayConfig) {
    this.config = config;
  }

  private trace(kind: Trace['kind'], text: string, detail?: string) {
    this.onTrace?.({ id: uuid(), kind, text, detail, at: Date.now() });
  }

  private setState(s: ConnectionState, error = '') {
    this.state = s;
    this.lastError = error;
    this.onState(s, error);
    if (s === 'connected') this.trace('system', 'Connected to gateway as ' + this.identity.nodeID);
    if (s === 'waitingForPairing') this.trace('system', 'Waiting for pairing approval…');
    if (s === 'error') this.trace('error', error || 'Connection error');
  }

  /** Send a user chat message into the session. */
  sendChat(text: string) {
    if (!text.trim()) return;
    this.trace('user', text);
    // Node→operator chat isn't part of the core v3 command set; send it as a
    // best-effort `chat` request so gateways that support it receive it.
    this.send({ type: 'req', id: uuid(), method: 'chat', params: { text } });
  }

  connect() {
    this.disconnect();
    this.setState('connecting');
    try {
      const ws = new WebSocket(configURL(this.config));
      this.ws = ws;
      ws.onmessage = (e) => this.handleMessage(String(e.data));
      ws.onerror = () => this.setState('error', 'Connection failed');
      ws.onclose = () => {
        if (this.state !== 'error') this.setState('disconnected');
      };
    } catch (e: any) {
      this.setState('error', e?.message ?? 'Invalid gateway URL');
    }
  }

  disconnect() {
    if (this.keepalive) clearInterval(this.keepalive);
    this.keepalive = null;
    this.ws?.close();
    this.ws = null;
    if (this.state !== 'error') this.setState('disconnected');
  }

  private handleMessage(data: string) {
    const frame = decodeFrame(data);
    if (!frame) return;
    if (frame.type === 'event') {
      if (frame.event === 'connect.challenge' && typeof frame.payload.nonce === 'string') {
        this.sendConnect(frame.payload.nonce);
      }
    } else if (frame.type === 'res') {
      if (frame.error) {
        this.setState(
          frame.error.code === 'NOT_PAIRED' ? 'waitingForPairing' : 'error',
          frame.error.message,
        );
      } else if (frame.ok && this.state === 'connecting') {
        this.setState('connected');
        this.startKeepalive();
      }
    } else if (frame.type === 'req') {
      this.dispatch(frame.id, frame.method, frame.params);
    }
  }

  private sendConnect(nonce: string) {
    const signedAt = Date.now();
    const signed = signedStringV3({
      deviceID: this.identity.deviceID,
      clientID: CLIENT_ID,
      clientMode: 'node',
      role: 'node',
      scopes: SCOPES,
      signedAtMs: signedAt,
      token: this.config.token,
      nonce,
      platform: 'ios',
      deviceFamily: 'rayban',
    });
    const signature = this.identity.sign(signed);
    this.send({
      type: 'req',
      id: uuid(),
      method: 'connect',
      params: {
        minProtocol: 3,
        maxProtocol: 4,
        client: { id: CLIENT_ID, mode: 'node', name: 'Super Meta' },
        role: 'node',
        scopes: SCOPES,
        caps: ['camera'],
        commands: ['camera.snap', 'camera.list', 'device.status', 'device.info'],
        auth: { token: this.config.token },
        device: {
          id: this.identity.deviceID,
          publicKey: this.identity.publicKeyB64Url,
          nonce,
          signedAt,
          signature,
          platform: 'ios',
          deviceFamily: 'rayban',
        },
      },
    });
  }

  private async dispatch(id: string, method: string, params: Record<string, any>) {
    if (method !== 'node.invoke') {
      this.respond(id, false, undefined, { code: 'UNKNOWN_METHOD', message: method });
      return;
    }
    const command = (params.command as string) ?? '';
    const args = (params.params as Record<string, any>) ?? {};
    this.trace('agent', `Agent ran ${command}`, Object.keys(args).length ? JSON.stringify(args) : undefined);
    try {
      const payload = await this.handler(command, args);
      this.trace('system', `${command} → ok`, summarize(payload));
      this.respond(id, true, payload);
    } catch (e: any) {
      const err: GatewayError =
        e && e.code ? e : { code: 'ERROR', message: e?.message ?? String(e) };
      this.trace('error', `${command} → ${err.code}`, err.message);
      this.respond(id, false, undefined, err);
    }
  }

  private startKeepalive() {
    if (this.keepalive) clearInterval(this.keepalive);
    this.keepalive = setInterval(() => {
      if (this.state === 'connected') this.send({ type: 'req', id: uuid(), method: 'ping', params: {} });
    }, 15000);
  }

  private respond(id: string, ok: boolean, payload?: Record<string, any>, error?: GatewayError) {
    this.send({ type: 'res', id, ok, payload, error });
  }

  private send(frame: Frame) {
    if (this.ws && this.ws.readyState === 1) this.ws.send(encodeFrame(frame));
  }
}

// Short one-line summary of a command result for the trace feed (base64 blobs elided).
function summarize(payload: Record<string, any>): string {
  const parts: string[] = [];
  for (const [k, v] of Object.entries(payload)) {
    if (k === 'base64' && typeof v === 'string') parts.push(`base64(${v.length}b)`);
    else parts.push(`${k}=${typeof v === 'object' ? JSON.stringify(v) : v}`);
  }
  return parts.join(' · ');
}

function uuid(): string {
  // RFC4122-ish v4 without crypto — ids are correlation-only, not security-sensitive.
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16);
  });
}
